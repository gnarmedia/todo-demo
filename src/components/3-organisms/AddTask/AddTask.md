```jsx
function handleSubmit(task) {
  const util = require('util');
  alert(`add task: ${util.inspect(task)}`);
  return false;
};
<AddTask onSubmit={(task) => handleSubmit(task)} />
```