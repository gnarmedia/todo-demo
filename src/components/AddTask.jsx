// import dependencies
import React, { Component } from "react";
import shortid from "shortid";

class AddTask extends Component {
    constructor(props) {
        super(props);
        this.state = { value: "" };

        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleChange(event) {
        this.setState({ value: event.target.value });
    }

    handleSubmit(event) {
        const task = {
            id: shortid.generate(),
            title: this.state.value,
            isComplete: false
        };

        this.props.onSubmit(task);
        this.setState((prevState, props) => ({
            value: ""
        }));

        event.preventDefault();
    }

    render() {
        return (
            <form onSubmit={this.handleSubmit}>
                <label>Add Task: </label>
                <input
                    type="text"
                    value={this.state.value}
                    onChange={this.handleChange}
                />
                <button type="submit">Add Task</button>
            </form>
        );
    }
}

// export our component for importing into other modules
export default AddTask;
