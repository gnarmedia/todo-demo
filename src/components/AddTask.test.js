import React from "react";
import { shallow } from "enzyme";

// import component
import AddTask from "./AddTask.jsx";

describe("AddTask", () => {
    it("should render the AddTask", () => {
        const wrapper = shallow(<AddTask />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should call onAddTaskSubmit when form is submitted", () => {
        const onAddTaskSubmitMock = jest.fn(),
            task = { id: "ryBRDzBdM", title: "Task 1", isComplete: false },
            wrapper = shallow(
                <AddTask onAddTaskSubmit={onAddTaskSubmitMock} />
            );

        // check assert default state
        expect(wrapper.state()).toMatchObject({ value: "" });
        expect(wrapper.find("button").exists()).toEqual(true);

        // assert existence of elements
        expect(wrapper.find("button").exists()).toEqual(true);
        expect(wrapper.find("input").exists()).toEqual(true);

        // set input value
        wrapper.find("input").simulate("change", { target: { value: "test" } });
        expect(wrapper.state()).toMatchObject({ value: "test" });

        // simulate submission and assert the call
        wrapper.instance().handleSubmit({ preventDefault: function() {} });
        expect(onAddTaskSubmitMock).toHaveBeenCalled();
    });
});
