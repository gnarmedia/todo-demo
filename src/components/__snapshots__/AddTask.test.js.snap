// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTask should render the AddTask 1`] = `
<form
  onSubmit={[Function]}
>
  <label>
    Add Task: 
  </label>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <button
    type="submit"
  >
    Add Task
  </button>
</form>
`;
